### PR TITLE
add dataset

### PR DESCRIPTION
**Pull request by andydiwenzhu**
## Related issue number

Fixes

### CODE REVIEW

File: python/vineyard/data/dataset.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+def dataset_builder(client, value, builder):
+    meta = ObjectMeta()
+    meta['typename'] = 'vineyard::DataSet'
+    meta['num'] = to_json(len(value))
+    for i in range(len(value)):
+        data, label = value[i]
+        meta.add_member(f'data_{i}_', builder.run(client, data))
+        meta.add_member(f'label_{i}_', builder.run(client, label))
+    return client.create_metadata(meta)
```
Comment (sighingnow) [inline python/vineyard/data/dataset.py:32]: How this dataset interactive with things like `DataFrame` or `Tensor`?

Shouldn't we build `DataSet` from `DataFrame` or `Tensor`, rather than create a new class? How could the new datatype interact with other applications bulid upon vineyard, like pandas as well as GAE?

File: python/vineyard/data/dataset.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+#! /usr/bin/env python
```
Comment (sighingnow) [inline python/vineyard/data/dataset.py:1]: Put the file to `python/vineyard/contrib` as `python/vineyard/contrib/pytorch.py`, as every ML framework has its own `DataSet`.

File: python/vineyard/data/dataset.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        self.resolver = resolver
+        self.num = from_json(obj.meta['num'])
+
+    def __len__(self):
+        return self.num
+
+    def __getitem__(self, idx):
+        data = self.resolver.run(self.obj.member(f'data_{idx}_'))
+        label = self.resolver.run(self.obj.member(f'label_{idx}_'))
+        return (data, label)
```
Comment (sighingnow) [inline python/vineyard/data/dataset.py:54]: `data` and `label` usually should have different number of columns?

File: python/vineyard/data/tests/test_dataset.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+class RandomDataset(Dataset):
+    def __init__(self, num):
+        self.num = num
+        self.ds = [(np.random.rand(4, 5, 6), np.random.rand(2, 3)) for i in range(num)]
+
+    def __len__(self):
+        return self.num
+
+    def __getitem__(self, idx):
+        return self.ds[idx]
```
Comment (sighingnow) [inline python/vineyard/data/tests/test_dataset.py:37]: I'm not very sure, but I think sampling from dataset in pytorch way should be use a `DataLoader`?

**Review by sighingnow — APPROVED** [review]

File: python/vineyard/data/dataset.py (hunk 1 above)
Comment (andydiwenzhu) [reply python/vineyard/data/dataset.py:32]: The Dataset type is indeed a composition type from dataframe and tensor. But the structure itself deserves a type.

File: python/vineyard/data/dataset.py (hunk 2 above)
Comment (andydiwenzhu) [reply python/vineyard/data/dataset.py:1]: Nope. I suppose this should be the "common" dataset type we would like to support in vineyard.

File: python/vineyard/data/dataset.py (hunk 3 above)
Comment (andydiwenzhu) [reply python/vineyard/data/dataset.py:54]: Here the shapes of data and label are independent.

File: python/vineyard/data/dataset.py (hunk 2 above)
Comment (sighingnow) [reply python/vineyard/data/dataset.py:1]: But it inherits from pytorch.

Pytorch itself shouldn't be dependent of vineyard.

File: python/vineyard/data/tests/test_dataset.py (hunk 4 above)
Comment (andydiwenzhu) [reply python/vineyard/data/tests/test_dataset.py:37]: Dataset is the input of DataLoader

File: python/vineyard/data/dataset.py (hunk 1 above)
Comment (sighingnow) [reply python/vineyard/data/dataset.py:32]: The how could map data generated by other engines to `Dataset` and shared with pytorch/tf/mxnet?

**Comment by sighingnow** [comment]
Close as we need reconcile the detailed design in #280.
